lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -1091,292 +544,461 @@ function Enable-Service {
-}
 
-function Disable-OVSExtension {
-    $ext = Get-OVSExtStatus
-    if ($ext -ne $null -and $ext.Enabled -eq $true) {
-        Disable-VMSwitchExtension $ovsExtName $ext.SwitchName
+    $optional = @{
+        "vhost"=$null;
+        "username"=$null;
+        "ha_queues"=$null;
     }
-    return $true
-}
 
-function Disable-OVS {
-    Stop-Service "ovs-vswitchd" -ErrorAction SilentlyContinue
-    Stop-Service "ovsdb-server" -ErrorAction SilentlyContinue
+    $ctx = Get-JujuRelationContext -Relation "amqp" -RequiredContext $required -OptionalContext $optional
 
-    Disable-Service "ovs-vswitchd"
-    Disable-Service "ovsdb-server"
+    $data = @{}
 
-    Disable-OVSExtension
+    if ($ctx.Count) {
+        if (!$ctx["username"]) {
+            $data["rabbit_userid"] = $username
+        } else {
+            $data["rabbit_userid"] = $ctx["username"]
+        }
+        if (!$ctx["vhost"]) {
+            $data["rabbit_virtual_host"] = $vhost
+        } else {
+            $data["rabbit_virtual_host"] = $ctx["vhost"]
+        }
+        if ($ctx["ha_queues"]) {
+            $data["rabbit_ha_queues"] = "True"
+        } else {
+            $data["rabbit_ha_queues"] = "False"
+        }
+        $data["rabbit_host"]=$ctx["hostname"];
+        $data["rabbit_password"]=$ctx["password"];
+    }
+    return $data
 }
 
-function Enable-OVS {
-    Enable-OVSExtension
+function Get-S2DContext {
+    [string]$instancesDir = (Get-JujuCharmConfig -Scope 'instances-dir').Replace('/', '\')
+    $ctxt = @{
+        "instances_path" = $instancesDir;
+        "compute_driver" = "hyperv.nova.driver.HyperVDriver";
+    }
 
-    Enable-Service "ovsdb-server"
-    Enable-Service "ovs-vswitchd"
+    $required = @{
+        "volumepath"=$null;
+    }
+    $s2dCtxt = Get-JujuRelationContext -Relation "s2d" -RequiredContext $required
+    $version = Get-JujuCharmConfig -Scope 'openstack-version'
+    $enableCluster = Get-JujuCharmConfig -Scope 'enable-cluster-driver'
+    if ($s2dCtxt.Count) {
+        if ($s2dCtxt["volumepath"] -and (Test-Path $s2dCtxt["volumepath"])) {
+            [string]$ctxt["instances_path"] = Join-Path $s2dCtxt["volumepath"] "Instances"
+        } else {
+            Write-JujuWarning "Relation information states that an s2d volume should be present, but could not be found locally."
+        }
+    }
+    if ($DISTRO_URLS[$version]['cluster'] -and $enableCluster) {
+        $ctxt['compute_driver'] = 'hyperv.nova.cluster.driver.HyperVClusterDriver'
+    }
 
-    Start-Service "ovsdb-server"
-    Start-Service "ovs-vswitchd"
+    # Try and create the instanced_dir on the clustered storage. We do not bother testing if the
+    # folder is already there before attempting to create it. There is a chance another node will
+    # create the folder between the Test-Path call and the mkdir call. Might as well try, and check
+    # for the existance of the folder if the command errors out.
+    try {
+        if (!(Test-Path $ctxt["instances_path"])) {
```

Comment:
Considering you are still testing the existence of the folder using test-path in the try block, please update the comment to reflect what happens here.

Something along the lines of:

```
# catch any IO error from mkdir, on the count that being a clustered storage, another node might create the folder between the time we Test-Path and the time we execute mkdir. Test again in case of IO exception.
```